### PR TITLE
New multiphase fluids package

### PR DESCRIPTION
**Pull request by jtclemm**
**Summary**

A new package RHEO (reproducing hydrodynamics and elastic objects) is added. It is an SPH fluid solver (with more modern numerical techniques than the SPH package). It couples to the BPM package to simulate fluid/solid systems. The package revolves around a new fix, fix rheo, as well as a few accessory fixes (to define constitutive equations for viscosity, pressure, etc.), computes, pair styles, and bonds styles. It also includes quite a few internal computes which are used to compartmentalize various calculations (e.g. the kernels and surface detection).

**Author(s)**

Joel Clemmer (SNL)
Thomas C. O’Connor (CMU)
Eric Palermo (CMU)

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

NA

**Implementation Notes**

While most changes are isolated to the RHEO and BPM packages, bond hybrid also had substantial changes to handle bonds that break. 
 
The package depends on GSL libraries for a few matrix operations in compute_rheo_kernel.cpp (maybe there’s another preferred numerical library). I tried set up cmake to handle this, but probably didn't succeed.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

### CODE REVIEW

**Comment by jtclemm** [comment]
With checks passing, the only task I have left is confirming the cmake system works. Ideally the RHEO package would always include the BPM package as a dependency, but admittedly I am not sure how to do that. I don't have much experience with cmake, so any assistance is appreciated.

**Comment by akohlmey** [comment]
> I don't have much experience with cmake, so any assistance is appreciated.

Please have a look at `cmake/CMakeLists.txt` lines 399 to 412.
You would just add a `pkg_depends(RHEO BPM)` and then  cmake will fail if BPM is not installed when RHEO is enabled.
A similar thing can be done with an `Install.sh` file in the `src/RHEO` folder for the GNU make build system.

File: src/BPM/bond_bpm.cpp [hunk 1]
```diff
@@ -82,7 +87,7 @@ BondBPM::~BondBPM()
   if (id_fix_dummy) modify->delete_fix(id_fix_dummy);
   if (id_fix_dummy2) modify->delete_fix(id_fix_dummy2);
   if (id_fix_update) modify->delete_fix(id_fix_update);
-  if (id_fix_bond_history) modify->delete_fix(id_fix_bond_history);
+  if (fix_bond_history) modify->delete_fix(id_fix_bond_history);
```
Comment (akohlmey) [inline src/BPM/bond_bpm.cpp:90]: Why this change?

**Review by akohlmey — CHANGES_REQUESTED** [review]
There doesn't seem to be any support for the traditional make based build system. It seems like there would need to be a `lib/rheo` folder with a Makefile.lammps and possibly an Install.py script (for auto detecting the GSL location and updating Makefile.lammps accordingly) that has the customization for the include and library paths to GSL. As already mentioned elsewhere, there also needs to be an `src/RHEO/Install.sh` script that realizes the dependency on BPM, but it would also need to incorporates the Makefile.lammps settings.

File: src/BPM/bond_bpm.cpp (hunk 1 above)
Comment (jtclemm) [reply src/BPM/bond_bpm.cpp:90]: Previously `id_fix_bond_history` was defined in the child class right before creating the fix. I decided to consolidate the fix initialization in the parent class and moved the id assignment to the child constructor to make it's definition more apparent. 

So now there's a gap between the id definition in the constructor and the fix definition in settings. This causes an error when reading a restart file since `modify` cannot find a fix with the given id.

File: src/RHEO/Install.sh [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+action fix_rheo_viscosity.h
+action fix_rheo_viscosity.cpp
+action fix_rheo.h
+action fix_rheo.cpp
+action pair_rheo.h
+action pair_rheo.cpp
+action pair_rheo_solid.h
+action pair_rheo_solid.cpp
+
+# package files with dependencies
```
Comment (akohlmey) [inline src/RHEO/Install.sh]: @jtclemm you don't need to do these "soft" dependencies (i.e. install only part of RHEO if BPM is not installed) if you also have a "hard" dependency (exit with an error when BPM is not installed) below.

You need to make a choice, which of the two you want. The hard dependency is obviously much easier to implement and makes some sense here. The soft dependency is more for packages like KSPACE or OPENMP.

If you want to have a soft dependency, then you need to have more CMake scripting to remove the files from RHEO that depend on sources in BPM. That one is rather complex to do (I can do it for you, if you want to go this route). The way I see it, there is little need for the effort (also for future expansions of either BPM or RHEO).

Just let me know your preferences. Since we are after the "stabilization deadline" there is not much of a rush, since this package would only be added after the stable release.

File: lib/rheo/Makefile.lammps [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# Settings that the LAMMPS build will import when this package is installed
```
Comment (akohlmey) [inline lib/rheo/Makefile.lammps:1]: This is technically correct, but it could be more automated using pkg-config or pkgconf.
Check out lib/kim, lib/kokkos, lib/qmmm

File: src/RHEO/Install.sh (hunk 2 above)
Comment (jtclemm) [reply src/RHEO/Install.sh]: That's a good point, I swapped it.

File: lib/rheo/Makefile.lammps (hunk 3 above)
Comment (jtclemm) [reply lib/rheo/Makefile.lammps:1]: I don't disagree with that! I did look at some of those libraries, but their automated solutions were out of my wheelhouse. I can take a stab at something more sophisticated, but it won't be for a while. I'll have to carve out time to work through some tutorials.

File: src/Depend.sh [hunk 4]
```diff
@@ -56,6 +56,10 @@ if (test $1 = "ASPHERE") then
   depend INTEL
 fi
 
+if (test $1 = "BPM") then
```
Comment (akohlmey) [inline src/Depend.sh]: This is not needed if you have a hard dependency. You need to uninstall RHEO first if you uninstall BPM or else you get an error.

File: lib/rheo/Makefile.lammps (hunk 3 above)
Comment (akohlmey) [reply lib/rheo/Makefile.lammps:1]: I can do it. I've done far more complex stuff in src/Makefile ... :wink:

Comment (akohlmey) [reply lib/rheo/Makefile.lammps:1]: Please have a look at my tweaks to the package. I also did some spell-checking...

Comment (jtclemm) [reply lib/rheo/Makefile.lammps:1]: That's really helpful and works great on my system. Thanks! I'll look through the changes in more detail and parse all the new additions.

File: src/atom.h [hunk 5]
```diff
@@ -155,6 +155,13 @@ class Atom : protected Pointers {
   double *eff_plastic_strain_rate;
   double *damage;
 
+  // RHEO package
+
+  int *status;
```
Comment (sjplimp) [inline src/atom.h]: this is a too-generic property name ?

File: doc/src/Howto_rheo.rst [hunk 6]
```diff
@@ -0,0 +1,115 @@
+Reproducing hydrodynamics and elastic objects (RHEO)
+====================================================
+
+The RHEO package is a hybrid implementation of smoothed particle
+hydrodynamics (SPH) for fluid flow, coupled to the :doc:`BPM package <Howto_bpm>`
```
Comment (sjplimp) [inline doc/src/Howto_rheo.rst]: Does this mean the RHEO package would never be used by itself?  Only in conjunction with BPM ?

File: doc/src/Howto_rheo.rst [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+to model solid elements. RHEO combines these methods to enable mesh-free modeling
+of multi-phase material systems. The SPH solver supports many advanced options
```
Comment (sjplimp) [inline doc/src/Howto_rheo.rst]: Its SPH solver ...

File: doc/src/Howto_rheo.rst [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+behavior. The package is designed with modularity in mind, so one can easily
+turn various features on/off, adjust physical details of the system, or
+develop new capabilities. For instance, the numerics associated with
+calculating gradients, reproducing kernels, etc. are separated into distinct
+classes to simplify the development of new integration schemes which can call
+these calculations. Additional numerical details can be found in
+:ref:`(Palermo) <howto_rheo_palermo>` and
+:ref:`(Clemmer) <howto_rheo_clemmer>`.
+
+Note, if you simply want to run a traditional SPH simulation, the SPH package
```
Comment (sjplimp) [inline doc/src/Howto_rheo.rst]: add link to Packages_details.rst for "SPH package"

Comment (sjplimp) [reply doc/src/Howto_rheo.rst]: Also, is RHEO  + BPM related at all to the SMD package (same author) which as I recall does something SPH-like for elastic solids.  If so, it would be good to also link to SMD here and highlight the differences.

File: src/atom.h (hunk 5 above)
Comment (jtclemm) [reply src/atom.h]: Yeah, probably. How about `rheo_status`?

File: doc/src/Howto_rheo.rst (hunk 6 above)
Comment (jtclemm) [reply doc/src/Howto_rheo.rst]: It can still be used by itself to run a pure-fluid simulation (e.g. the Taylor-Green example). But it still needs to be built with BPM.

File: doc/src/Howto_rheo.rst (hunk 8 above)
Comment (jtclemm) [reply doc/src/Howto_rheo.rst]: My understanding of SMD is minimal, but I can add a mention.

File: src/atom.h (hunk 5 above)
Comment (sjplimp) [reply src/atom.h]: I like it

Comment (jtclemm) [reply src/atom.h]: Done

File: doc/src/Howto_rheo.rst (hunk 6 above)
Comment (jtclemm) [reply doc/src/Howto_rheo.rst]: I clarified the language

File: doc/src/Howto_rheo.rst (hunk 7 above)
Comment (jtclemm) [reply doc/src/Howto_rheo.rst]: Done

File: doc/src/Howto_rheo.rst (hunk 8 above)
Comment (jtclemm) [reply doc/src/Howto_rheo.rst]: Added links to both packages.

File: src/bond_hybrid.cpp [hunk 9]
```diff
@@ -135,6 +142,19 @@ void BondHybrid::compute(int eflag, int vflag)
     }
   }
 
+  // if bond type can be set to zero and deleted, update bondlist_orig
```
Comment (sjplimp) [inline src/bond_hybrid.cpp]: can you make this a more descriptive (few line) comment about the need for orig_map and the issue it is solving - i.e. to propagate a broken bond from the temporary list to original list, so that bond is truly broken, else it could "reappear" when neigh lists are rebuilt

Comment (jtclemm) [reply src/bond_hybrid.cpp]: Yes, that sounds like a good idea. Do you think the new comment is sufficiently clear?

**Review by sjplimp — APPROVED** [review]

File: src/bond_hybrid.cpp (hunk 9 above)
Comment (sjplimp) [reply src/bond_hybrid.cpp]: yes, looks good - I just approved the PR

**Comment by jtclemm** [comment]
@sjplimp Thanks for reviewing the changes. I also have a few demo videos which can be added to the website if relevant.



https://github.com/user-attachments/assets/9fc2e4ec-f016-4856-a6ce-8643b9bcf7db

https://github.com/user-attachments/assets/93d3e7d8-5159-4713-bc83-0a6509c6dfd4

https://github.com/user-attachments/assets/7dbf9771-0162-458f-a37f-1ac344c01c48

**Review by akohlmey — APPROVED** [review]
